File: 06_Strings/Valid_Palindrome_II/valid_palindrome_II.cpp
// LeetCode Problem 680 - https://leetcode.com/problems/valid-palindrome-ii/submissions/1905210934/
// TC - O(n), SC - O(1)

bool isp(string s, int l, int r) {
        for (int i = l, j = r; i < j; i++, j--)
            if (s[i] != s[j]) return false;
        return true;
    }
    bool validPalindrome(string s) {
        // for (int i = 0, j = s.size() - 1; i < j; i++, j--)
        //     if (s[i] != s[j]) return isp(s, i + 1, j) || isp(s, i, j - 1);
        // return true;
        int start  = 0, end = s.size()-1;
        while(start <= end){
            if (s[start] != s[end]) return isp(s, start + 1, end) || isp(s, start, end - 1);
            start++;
            end--;
        }
        return true;
    }

File: 06_Strings/Valid_Palindrome_II/valid_palindrome_II.java
// LeetCode Problem 680 - https://leetcode.com/problems/valid-palindrome-ii/submissions/1905210934/
// TC - O(n), SC - O(1)

public class valid_palindrome_II {
  
    private boolean isPal(String s, int l, int r) {
        while (l < r) {
            if (s.charAt(l) != s.charAt(r)) return false;
            l++;
            r--;
        }
        return true;
    }

    // Main function
    public boolean validPalindrome(String s) {
        int start = 0, end = s.length() - 1;

        while (start < end) {
            if (s.charAt(start) != s.charAt(end)) {
                // Skip either left or right character once
                return isPal(s, start + 1, end) || isPal(s, start, end - 1);
            }
            start++;
            end--;
        }
        return true;
    }
}
